content/js3/blocks/asynchrony/index.md

```diff
@@ -0,0 +1,71 @@
++++
+title = '⏳ Asynchrony : outside time'
+headless = true
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+1="Define asynchrony"
+2="Explain why we need asynchrony"
+3="Identify an asynchronous method we have already used"
++++
+
+We can handle latency using {{<tooltip title="asynchronous execution">}}run code in a different order.{{</tooltip>}} To understand asynchrony we first need to be clear about {{<tooltip title="synchronous execution">}}run code in the order it is written.{{</tooltip>}}.
+
+We have written a lot of JavaScript programs that execute sequentially. This means that each line of code is run in order, one after the other.
+{{<columns>}}
+
+#### For example:
+
+```js
+console.log("first");
+console.log("second");
+console.log("third");
+```
+
+<--->
+
+#### Outputs:
+
+```console
+first
+second
+third
+```
+
+{{</columns>}}
+Each line of code is run in order. This is synchronous execution. We do this because JavaScript is {{<tooltip title="single threaded">}}
+A single thread can do one thing at a time. JavaScript is a single threaded language.
+{{</tooltip>}}.
+
+When we call a function, the function will run to completion before the next line of code is executed. But what if we need to wait for something to happen? What if we need to wait for our data to arrive before we can show it? In this case, we can use asynchronous execution.
+
+{{<tabs name="Event Loop">}}
+{{<tab name="Event Listener">}}
+We have already used asynchronous execution. We have defined eventListeners that _listen_ for events to happen, _then_ execute a callback function. But here's a new idea: eventListeners are part of the [Event API](https://developer.mozilla.org/en-US/docs/Web/API/Event). They are not part of JavaScript! 🤯
+
+When you set an eventListener you are really sending a call to a Web API and asking it do something for you.
+
+```js
+const search = document.getElementById("search");
+const waitForEvent = search.addEventListener("input", handleInput);
+```
+
+The callback `handleInput` cannot run until the user types. With `fetch`, the callback function cannot run until the data arrives. In both cases, we are waiting for something to happen before we can run our code.
+{{</tab>}}
+{{<tab name="Visualise the Event Loop">}}
+
+<iframe src="http://latentflip.com/loupe/?code=JC5vbignYnV0dG9uJywgJ2NsaWNrJywgZnVuY3Rpb24gb25DbGljaygpIHsKICAgIGNvbnNvbGUubG9nKCdZb3UgY2xpY2tlZCB0aGUgYnV0dG9uIScpOyAgICAKfSk7Cgpjb25zb2xlLmxvZygiSGkhIik7Cgpjb25zb2xlLmxvZygiV2VsY29tZSB0byB0aGUgZXZlbnQgbG9vcCIpOw%3D%3D!!!PGJ1dHRvbj5DbGljayBtZSE8L2J1dHRvbj4%3D" width="100%" height="480px"></iframe>
```

Comment:
Are we expecting the trainees to watch the 27 minute video in the loupe help text?

I haven't watched it yet, but assuming it's good, I think we probably should (and should tell them to). If not, I think we need to explain "call stack" and "callback queue" a bit before people can find this super enlightening?

Either way, I think this block is more than 30 minutes.

Reply:
Yes please update the time!

I like the video a lot and have used it with trainees. I will say that about half of them liked it and half of them preferred just using the loupe, and only really got the video after playing with the visualiser. This is pretty classic -- some people are principles first learners and some are practical first.